b2/progress.py

```diff
@@ -218,10 +218,12 @@ def read(self, size=None):
         return data
 
 
-class StreamWithProgress(object):
+class AbstractStreamWithProgress(object):
     """
     Wraps a file-like object and updates a ProgressListener
-    as data is read and written.
+    as data is read / written.
+    In the abstract class, read and write methods do not update
```

Comment:
Breaking out read and write streams is an improvement.  Now that it's clear that some streams are for reading and some for writing, it raises the question whether reading streams need a `write` method, and whether writing streams need a `read` method.   If not, it would clarify things even more.

Reply:
They do:
in the new downloading implementation, the writing stream does some reading after everything is written, to complete hashing which has to be done start to end.

As they are now, progress listener wrappers can be chained to see both reading and writing from the stream separately. I would not like to take that away.